kernel/src/net/socket/ip/stream/init.rs

```diff
@@ -90,19 +147,66 @@ impl InitStream {
 
         match ListenStream::new(bound_port, backlog, option, observer) {
             Ok(listen_stream) => Ok(listen_stream),
-            Err((bound_port, error)) => Err((error, Self::Bound(bound_port))),
+            Err((bound_port, error)) => Err((error, Self::new_bound(bound_port))),
         }
     }
 
-    pub fn local_endpoint(&self) -> Option<IpEndpoint> {
-        match self {
-            InitStream::Unbound => None,
-            InitStream::Bound(bound_port) => Some(bound_port.endpoint().unwrap()),
+    pub fn try_recv(
+        &self,
+        _writer: &mut dyn MultiWrite,
+        _flags: SendRecvFlags,
+    ) -> Result<(usize, SocketAddr)> {
+        // FIXME: Linux does not return addresses for `recvfrom` on connection-oriented sockets.
+        // This is a placeholder that has no Linux equivalent. (Note also that in this case
+        // `getpeeraddr` will simply fail with `ENOTCONN`).
+        const UNSPECIFIED_SOCKET_ADDR: SocketAddr = SocketAddr::IPv4(Ipv4Addr::UNSPECIFIED, 0);
+
+        // Below are some magic checks to make our behavior identical to Linux.
+
+        if self.can_connect_new {
+            return_errno_with_message!(Errno::ENOTCONN, "the socket is not connected");
+        }
+
+        if let Some(err) = self.test_and_clear_error() {
+            return Err(err);
         }
+
+        Ok((0, UNSPECIFIED_SOCKET_ADDR))
+    }
+
+    pub fn try_send(&self, _reader: &mut dyn MultiRead, _flags: SendRecvFlags) -> Result<usize> {
+        if let Some(err) = self.test_and_clear_error() {
+            return Err(err);
+        }
+
+        return_errno_with_message!(Errno::EPIPE, "the socket is not connected");
+    }
+
+    pub fn local_endpoint(&self) -> Option<IpEndpoint> {
+        self.bound_port
+            .as_ref()
+            .map(|bound_port| bound_port.endpoint().unwrap())
     }
 
     pub(super) fn check_io_events(&self) -> IoEvents {
         // Linux adds OUT and HUP events for a newly created socket
-        IoEvents::OUT | IoEvents::HUP
+        let mut events = IoEvents::OUT | IoEvents::HUP;
+
+        if self.is_conn_refused.load(Ordering::Relaxed) {
+            events |= IoEvents::ERR;
```

Comment:
I haven't noticed anywhere that calls `self.pollee.notify(IoEvents::ERR)` when the connection is refused. But from this code, `ERR` should be notified if the connection is refused. Did I miss something?

Reply:
It's notified by `SocketEvent::CLOSED`. I've added `IoEvents::ERR` below:
https://github.com/asterinas/asterinas/blob/794327616ca24483c8cd4d2f38b69ca963da0094/kernel/src/net/socket/ip/stream/observer.rs#L32-L35

Reply:
Got it! Thanks for your explanation!